### ISSUE TITLE
Display map of heap is not showing

### ISSUE THREAD

**Issue by azrielpamungkas** [radareorg/radare2#21762]
## Environment

```
Mon May 22 08:29:00 AM 2023
radare2 5.5.0 0 @ linux-x86-64 git.5.5.0
Linux x86_64
```

When I want to check the appearance of the map of heap a warning appears like this
*Warning: Can't find glibc mapped in memory (see dm)*

> [0x7fad1cf049c0]> pdf @ main
            ; DATA XREF from entry0 @ 0x4010a4
┌ 188: int main (int argc, char **argv);
│           ; var int64_t var_20h @ rbp-0x20
│           ; var int64_t var_14h @ rbp-0x14
│           ; var int64_t var_10h @ rbp-0x10
│           ; var int64_t var_8h @ rbp-0x8
│           ; arg int argc @ rdi
│           ; arg char **argv @ rsi
│           0x004011a2      55             push rbp
│           0x004011a3      4889e5         mov rbp, rsp
│           0x004011a6      4883ec20       sub rsp, 0x20
│           0x004011aa      897dec         mov dword [var_14h], edi    ; argc
│           0x004011ad      488975e0       mov qword [var_20h], rsi    ; argv
│           0x004011b1      837dec01       cmp dword [var_14h], 1
│       ┌─< 0x004011b5      7f19           jg 0x4011d0
│       │   0x004011b7      488d05c20e00.  lea rax, str.Please_specify_an_argument_to_copy_:__ ; 0x402080 ; "Please specify an argument to copy :-)"
│       │   0x004011be      4889c7         mov rdi, rax
│       │   0x004011c1      e87afeffff     call sym.imp.puts           ; int puts(const char *s)
│       │   0x004011c6      bf01000000     mov edi, 1
│       │   0x004011cb      e8b0feffff     call sym.imp.exit
│       └─> 0x004011d0      bf40000000     mov edi, 0x40               ; elf_phdr
│           0x004011d5      e886feffff     call sym.imp.malloc         ;  void *malloc(size_t size)
│           0x004011da      488945f8       mov qword [var_8h], rax
│           0x004011de      bf40000000     mov edi, 0x40               ; elf_phdr
│           0x004011e3      e878feffff     call sym.imp.malloc         ;  void *malloc(size_t size)
│           0x004011e8      488945f0       mov qword [var_10h], rax
│           0x004011ec      488b45f0       mov rax, qword [var_10h]
│           0x004011f0      488d1595ffff.  lea rdx, [sym.nowinner]     ; 0x40118c
│           0x004011f7      488910         mov qword [rax], rdx
│           0x004011fa      488b45e0       mov rax, qword [var_20h]
│           0x004011fe      4883c008       add rax, 8
│           0x00401202      488b10         mov rdx, qword [rax]
│           0x00401205      488b45f8       mov rax, qword [var_8h]
│           0x00401209      4889d6         mov rsi, rdx
│           0x0040120c      4889c7         mov rdi, rax
│           0x0040120f      e81cfeffff     call sym.imp.strcpy         ; char *strcpy(char *dest, const char *src)
│           0x00401214      488b45f0       mov rax, qword [var_10h]
│           0x00401218      488b08         mov rcx, qword [rax]
│           0x0040121b      488b55f0       mov rdx, qword [var_10h]
│           0x0040121f      488b45f8       mov rax, qword [var_8h]
│           0x00401223      4889c6         mov rsi, rax
│           0x00401226      488d057b0e00.  lea rax, str.data_is_at__p__fp_is_at__p__will_be_calling__p_n ; 0x4020a8 ; "data is at %p, fp is at %p, will be calling %p\n"
│           0x0040122d      4889c7         mov rdi, rax
│           0x00401230      b800000000     mov eax, 0
│           0x00401235      e816feffff     call sym.imp.printf         ; int printf(const char *format)
│           0x0040123a      488b05ff2d00.  mov rax, qword [obj.stdout] ; obj.__TMC_END__
│                                                                      ; [0x404040:8]=0
│           0x00401241      4889c7         mov rdi, rax
│           0x00401244      e827feffff     call sym.imp.fflush         ; int fflush(FILE *stream)
│           0x00401249      488b45f0       mov rax, qword [var_10h]
│           0x0040124d      488b10         mov rdx, qword [rax]
│           0x00401250      b800000000     mov eax, 0
│           0x00401255      ffd2           call rdx
│           0x00401257      b800000000     mov eax, 0
│           0x0040125c      c9             leave
└           0x0040125d      c3             ret
[0x7fad1cf049c0]> db 0x00401214
[0x7fad1cf049c0]> dc
hit breakpoint at: 0x401214
[0x00401214]> dmh
Warning: Can't find glibc mapped in memory (see dm)
[0x00401214]> dm
0x0000000000400000 - 0x0000000000401000 - usr     4K s r-- /home/sebastian/Downloads/heap0/pwn-32 /home/sebastian/Downloads/heap0/pwn-32 ; segment.ehdr
0x0000000000401000 - 0x0000000000402000 * usr     4K s r-x /home/sebastian/Downloads/heap0/pwn-32 /home/sebastian/Downloads/heap0/pwn-32 ; map._home_sebastian_Downloads_heap0_pwn_32.r_x
0x0000000000402000 - 0x0000000000403000 - usr     4K s r-- /home/sebastian/Downloads/heap0/pwn-32 /home/sebastian/Downloads/heap0/pwn-32 ; map._home_sebastian_Downloads_heap0_pwn_32.r__
0x0000000000403000 - 0x0000000000404000 - usr     4K s r-- /home/sebastian/Downloads/heap0/pwn-32 /home/sebastian/Downloads/heap0/pwn-32 ; map._home_sebastian_Downloads_heap0_pwn_32.rw_
0x0000000000404000 - 0x0000000000405000 - usr     4K s rw- /home/sebastian/Downloads/heap0/pwn-32 /home/sebastian/Downloads/heap0/pwn-32 ; reloc.strcpy
0x000000000135d000 - 0x000000000137e000 - usr   132K s rw- [heap] [heap]
0x00007fad1cced000 - 0x00007fad1ccf0000 - usr    12K s rw- unk0 unk0
0x00007fad1ccf0000 - 0x00007fad1cd16000 - usr   152K s r-- /usr/lib/x86_64-linux-gnu/libc.so.6 /usr/lib/x86_64-linux-gnu/libc.so.6
0x00007fad1cd16000 - 0x00007fad1ce6b000 - usr   1.3M s r-x /usr/lib/x86_64-linux-gnu/libc.so.6 /usr/lib/x86_64-linux-gnu/libc.so.6
0x00007fad1ce6b000 - 0x00007fad1cebe000 - usr   332K s r-- /usr/lib/x86_64-linux-gnu/libc.so.6 /usr/lib/x86_64-linux-gnu/libc.so.6
0x00007fad1cebe000 - 0x00007fad1cec2000 - usr    16K s r-- /usr/lib/x86_64-linux-gnu/libc.so.6 /usr/lib/x86_64-linux-gnu/libc.so.6
0x00007fad1cec2000 - 0x00007fad1cec4000 - usr     8K s rw- /usr/lib/x86_64-linux-gnu/libc.so.6 /usr/lib/x86_64-linux-gnu/libc.so.6
0x00007fad1cec4000 - 0x00007fad1ced1000 - usr    52K s rw- unk1 unk1
0x00007fad1cee8000 - 0x00007fad1ceea000 - usr     8K s rw- unk2 unk2
0x00007fad1ceea000 - 0x00007fad1ceeb000 - usr     4K s r-- /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
0x00007fad1ceeb000 - 0x00007fad1cf10000 - usr   148K s r-x /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 ; map._usr_lib_x86_64_linux_gnu_ld_linux_x86_64.so.2.r_x
0x00007fad1cf10000 - 0x00007fad1cf1a000 - usr    40K s r-- /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 ; map._usr_lib_x86_64_linux_gnu_ld_linux_x86_64.so.2.r__
0x00007fad1cf1a000 - 0x00007fad1cf1c000 - usr     8K s r-- /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 ; map._usr_lib_x86_64_linux_gnu_ld_linux_x86_64.so.2.rw_
0x00007fad1cf1c000 - 0x00007fad1cf1e000 - usr     8K s rw- /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
0x00007fffb184c000 - 0x00007fffb186d000 - usr   132K s rw- [stack] [stack] ; map._stack_.rw_
0x00007fffb1920000 - 0x00007fffb1924000 - usr    16K s r-- [vvar] [vvar] ; map._vvar_.r__
0x00007fffb1924000 - 0x00007fffb1926000 - usr     8K s r-x [vdso] [vdso] ; map._vdso_.r_x
[0x00401214]>

**Comment by gogo2464** [comment]
Yes, I agree with pancake. Do `git clone https://github.com/radareorg/radare2; cd radare2; sh sys/install.sh`

then try with:

`dmhg `dmha~main_arena[2]`

Then could you tell us if you still have a bug please?

**Comment by gogo2464** [comment]
@trufae probably a package from a Linux distro.

**Comment by azrielpamungkas** [comment]
> Yes, I agree with pancake. Do `git clone https://github.com/radareorg/radare2; cd radare2; sh sys/install.sh`
> 
> then try with:
> 
> ```
> dmhg `dmha~main_arena[2]`
> ```
> 
> Then could you tell us if you still have a bug please?

I don't know if this is still a bug or something is wrong with me, let's compare the results with this
![Screenshot_2023-05-23_21_37_35](https://github.com/radareorg/radare2/assets/100214154/71bbeacb-a8a5-41ed-9a00-c81eee5d8c4a)

**Comment by radare** [comment]
Is this output with the latest r2? anyway it's been like 5 years since someone had a look at this code, and libc changes the heap algorithm every 1-2 years, so its expected to break, shuold be easy to fix but i cant handle more unrelated paralel development right now, if you want to dig into the code to fix it i can guide you

**Comment by azrielpamungkas** [comment]
yes, the output is from the latest r2. Please provide me with a simple guide to fix it, and I will do my best to resolve the issue.

**Comment by radare** [comment]
(sorry for the delay, i was busy re-reading the code and cleaning it up a bit for giving you an explanation)

Cool! Thanks for confirming on master. Let me give you some pointers in the source to locate all the code of interest:

You can find the structs used by the heap allocator from gnu libc in:
* https://github.com/bminor/glibc/blob/glibc-2.28/malloc/malloc.c#L1658

In r2, the dmh commands are handled in libr/core/cmd_debug.c you can /search for `"dmh"` (with the quotes included to find the code that implements the handler. (spoiler: it's `r_debug_heap`). which is implemented in the same file (despite it should be moved to r_debug at some point.

* This function switches between glibc and jemalloc (or windows). But i have added initial macos support now.
* You can check the code that the bug should be in dmh_glibc.c

This code is glibc specific, but i would like to make it generic, but didnt wanted to put any pressure to the contributors at the time, i will do the refactoring when everything is in shape again, for now i would say that its important to get it working for latests glibc, and maybe add tests, but this is hard because requires shipping old linuxes on the CI and so on.

anyway, the dmh command does the following

* find glibc in memory (uses dm - debugger maps) to find the base address)
* runs rabin2 (using the api) on the associated file from disk to find the symbols
* when the global heap arena symbol is resolved then it walks the pointers
* the GHT() macro allows to implement the same code for 32 and 64bit

So having this in mind

* use eprintf statements to follow the calls and trace the data
* create simple programs and oneliners with r2 for quickly testing
* read code from GEF and GLIBC to follow the structs

Notes for my future self:

* The macOS allocator support zones, which is multiple isolated heap arenas for a single program
* ^ i will take this as inspiration for the generic api in the upcoming release to unify all those implementations
* this code should be independent from the debugger, because it must work on static or with remote debugging
* standarizing with common apis/structs will make it easier to export to JSON and unify the output for all heap implementations

**Comment by trufae** [comment]
any other question?

**Comment by gogo2464** [comment]
me? I was waiting as well lol.

**Comment by koyaan** [comment]
@gogo2464  Please try again from git if still bugged, provide binary + glibc version used
